pkg/operator/controller/typed.go

```diff
@@ -0,0 +1,193 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type TypedBuilder interface {
+	Builder
+
+	// For updates the builder to watch the client.Object passed in
+	For(client.Object) TypedBuilder
+}
+
+type TypedBuilderControllerRuntimeAdapter struct {
+	builder *controllerruntime.Builder
+}
+
+func NewTypedBuilderControllerRuntimeAdapter(builder *controllerruntime.Builder) *TypedBuilderControllerRuntimeAdapter {
+	return &TypedBuilderControllerRuntimeAdapter{
+		builder: builder,
+	}
+}
+
+func (t *TypedBuilderControllerRuntimeAdapter) For(obj client.Object) TypedBuilder {
+	t.builder = t.builder.For(obj)
+	return t
+}
+
+func (t *TypedBuilderControllerRuntimeAdapter) Complete(r reconcile.Reconciler) error {
+	return t.builder.Complete(r)
+}
+
+type TypedReconciler[T client.Object] interface {
+	Reconcile(context.Context, T) (T, reconcile.Result, error)
```

Comment:
I think we should just rely on mutation of T. This pattern is used broadly in k8s controllers and I don't see a great reason to deviate. Using the pointer parameter style makes it explicit that this function is intended to mutate. With both mechanisms, it's not clear the responsibilities of the input provisioner or output provisioner. 
```suggestion
	Reconcile(context.Context, T) (reconcile.Result, error)
```

Reply:
Let's talk about this more. In general, I don't like impure functions. I think it makes it ambiguous to developers that are writing the code that the object they are mutating is going to affect the etcd object. Additionally, it doesn't allow users to opt-out to mutation by returning nil like returning the object would.

Reply:
Impure functions can have a performance impact, and should be used strategically and clearly. Generally, I agree with you, but not for controllers.